### ISSUE TITLE
write a kubetest deployer for kind

### ISSUE THREAD

**Issue by neolit123** [kubernetes-sigs/kind#161]
@BenTheElder not sure if a tracking issue already exists for this (tried searching for it).
i can try helping with the initial work for a deployer.

/assign
/kind feature
/priority important-soon

**Comment by BenTheElder** [comment]
Thanks, one did not and that would be great!

I had something sketched out but never got it done. Two ideas from that:

- fetch a binary release for the platform and stick it in `$HOME/.kubetest/kind/$binary` so we can call it from kubetest.
  - checksum (sha256) of the binary there would let us avoid re-downloading when it is already downloaded. We could pre-download in the kubetest CI image too once this works.
- we'll probably need to integrate with --build too so it does a kind build node-image instead. We will probably have to make build depend on the deployer 🤷‍♂️

**Comment by fabriziopandini** [comment]
@neolit123 @BenTheElder 
Might be a silly question, but are there some documents about how kubetest could be configured / how kubetest parameters will flow down to kind?

I ask this because IMO those info are the key requirements for designing kind UX for CI use case; more specifically I'm thinking at:
- how can I specify cluster Config/a desired cluster topology
- how can I specify a complex test plan like e.g init-join-e2e and then upgrade-join another node-e2e again etc

**Comment by neolit123** [comment]
> Might be a silly question, but are there some documents about how kubetest could be configured / how kubetest parameters will flow down to kind?

i don't know if such a document exists.

looking at my experience with k-a, we have to store the config scenarios "somewhere" as we cannot expose the whole config or actions as CLI flags from jobs to be passed to the kind kubetest deployer.

in the case of k-a the configs were stored in the k-a repo and we were binding test-infra (kubeadm image) to a specific version of k-a. in my view this is problematic because any tiny change we make in the config scenarios has to be reflected in test-infra and images have to be rebuild.

that is the reason i pushed hard for the creation of the k-a `kubeadm-e2e` branch. this removed the need for us to hardcode a specific SHA and made it possible for use to use the tip of a branch.

so my proposal for kind is to use something similar. it's just more flexible.

**Comment by neolit123** [comment]
@BenTheElder 
> For parameters flowing down, we can just plumb through a few like the kind config path

i started implementing the deployer interface yesterday and this is pretty much the only flag for now.

i have a number of questions, but the few major ones are:
1) where to store the configs?
2) would the config allow actions and test scenarios (cc @fabriziopandini ) my assumption is that this is our plan?
3) i need a tutorial for testing the deployer locally. i can build the new kubetest and run it inside a container, but if i want to test the actual periodic jobs YAMLs this would mean that kubekins has to be rebuilt and from there i'm kind of lost with the bootstrap.py part.

some options for 1):
- should we stored them in a folder under kind/master, this will allow `--repo` to pick them up or do we want to use tags and branches for the repo?
- should we also allow configs to be fetched by the kubetest deployer from http:// , not sure if the pods that run jobs are sandboxed in that regard?

for 2) we can do this in time and the deployer can be created in parallel with the minimal single node master test case.

**Comment by neolit123** [comment]
re: `test-cmd`:

to me it seems that `test-cmd` can be considered as an alternative to a deployer:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-periodic-jobs.yaml#L18-L42
in the above `scenario=kubernetes_e2e` case it is used to call a bash script.

the current kind periodics on the other hand do `scenario=execute` and also pass a bash script as an argument at the end:
https://github.com/kubernetes/test-infra/blob/26741576567fbb7abcaa5f71b05e3bdbd6f67c06/config/jobs/kubernetes-sigs/kind/kind.yaml#L97

to my understanding we want to write the deployer so that we don't want to call scripts?

**Comment by fabriziopandini** [comment]
@neolit123 there is no plan defined yet
my personal opinion is that actions should be kept separated by config, so you can combine actions-plans and cluster topologies (vs being forced to duplicate one of the two if both conflate in the same config file).
If this will be the choice that gets move consensus, I assume that  actions-plans will be  provided by dedicated flags like e.g. `--action []string` and eventually also `--action-plan yaml`

**Comment by BenTheElder** [comment]
forgot to mention, there is a rough MVP merged now (thanks!) but we need to follow up with:
- using a stable release binary
- proper support for kubernetes from source
- plumbing through config

**Comment by neolit123** [comment]
> using a stable release binary

how should we do that?

do we simply need to always download the latest kind amd64 binary for linux from github?
or should we expose a "kind-release-tag" flag to the deployer to support?

e.g.
`--kind-release-tag=latest` means latest release tag.
`--kind-release-tag=0.1.0` means a fixed tag.

> proper support for kubernetes from source

we already can clone k/k per job, so what else is needed to satisfy "kubernetes from source"?
my understanding was that we might want to expose a way to use the pre-built node image from a bucket?

> plumbing through config

the kind config can be passed per job.
can you please elaborate what has to be plumbed.

**Comment by BenTheElder** [comment]
> how should we do that?

I think the sanest and most performant thing for now is something like:
- use a fixed tag, PR to kubetest to upgrade
- download the binary at that tag that matches the host OS
- checksum the binary
- on future runs detect the downloaded binary and match the expected checksum, if it doesn't match then download again

We can stick it in something like `$HOME/.kubetest/kind/kind-$OS-amd64`

> my understanding was that we might want to expose a way to use the pre-built node image from a bucket?

for periodics maybe, for k/k presubmit in particular we need to build a node image from kubetest, and specify that kind run that image


> the kind config can be passed per job.
> can you please elaborate what has to be plumbed.

I might be forgetting, but I don't think we pass through a config flag from kubetest to kind yet.

**Comment by BenTheElder** [comment]
> but this would require rebuilding kubekins, correct?

yes, but only when we update.

> so if we download it from a prow test job the first time and store it in that location, how would it persist the next time the same job is ran? is there some sort of caching?

locally it will cache, for CI we can rig something where we already obtain that binary and stick it in the expected location.

> the deployer already supports that. but only that - i.e. always build the node image and always use that in kind create. :)

ah, we should do this in --build instead of --up though probably

> already supported:

thanks!

**Comment by neolit123** [comment]
> yes, but only when we update.

ack.

> locally it will cache, for CI we can rig something where we already obtain that binary and stick it in the expected location.

ack.

> ah, we should do this in --build instead of --up though probably

this is tricky. --build accepts only certain [values](https://github.com/kubernetes/test-infra/blob/aa1397ee7a4bd0b322309433b422f30976a2a7fa/kubetest/build.go#L57) and there is not way to plug `kind build` into that, as far as i can tell.

perhaps this is worth of adding custom flags to the kind deployer only:

`--kind-build-note-image`, where if passed it will build based on the fetched k/k `--repo` and branch. otherwise it will download (+cache) the pre-built node image.

`--kind-build-binary` that will build the kind binary otherwise it will download (+cache) the release binary from github.

WDYT?

**Comment by BenTheElder** [comment]
> this is tricky. --build accepts only certain values and there is no way to plug kind build into that, as far as i can tell.

we could map --build=bazel -> --type=bazel, --build=quick-release -> default kind build (make/bash/docker), and just not support apt on the kind end, or values that kind doesn't know about.
we'd only do this when the kind deployer flag was set or something... 🤔 

CI jobs generally stick to those two options.

it may be ugly to hook, but that's actually one of the things that will help drive a better kubetest refactor later is understanding what things we need to hook. we're not the only ones with extra build steps

> --kind-build-binary that will build the kind binary otherwise it will download (+cache) the release binary from github.

that one i've been thinking about too and don't have a good answer, maybe let's start with getting the stable release binary and come back to how to use an alternate binary later ... (maybe even a $KIND env to override 😱)

---

sorry for the delay - did not expect so many fires today, especially not prow being down this morning and writing that up. thanks again :-)

**Comment by neolit123** [comment]
sending update that fixes some TODOs in the kind deployer later today.

**Comment by BenTheElder** [comment]
We still have plenty of room to iterate on this, but I'm happy to call this done for now, thanks for the PRs @neolit123!

**Comment by BenTheElder** [comment]
We should document this somewhere, not sure where yet, will open another issue here to track for now, but possibly those docs should live eg in the upstream e2e test docs, or test-infra.
